### ISSUE TITLE
Brainstorm & Document benefit module function signatures

### ISSUE THREAD

**Issue by isaacchansky** [codeforboston/project-hope#12]
All of these modules should have some sort of standard parameter & result signature. 

i.e. :
```
// bigOlParameterObject looks like:
// {
//    'householdIncome': <Float>,
//    'numberOfChildren': <Integer>,
//    ...
// }
function someBenefitCalc(bigOlParameterObject){
  ...
  return {
    eligible: true,
    value: 1000
  }
}
```

Let's discuss!

**Comment by vnwhlr** [comment]
Some of these functions will take a _ton_ of arguments, so it might be easier to just pass the whole object representing the client in. We can destructure it like this in ES6: https://simonsmith.io/destructuring-objects-as-function-parameters-in-es6/

As far as return values go: I think in addition to the "eligible" and "value" fields will work, we may want to consider a "reason" or "reasons". My thinking is that eventually we may have to return "risk factors" computed from how close a client is to the thresholds for eligibility - for example, if you're eligible if your gross income is $10,000, and your gross income is $8,000, we might want to return something that indicates that if your income rises by more than $2,000, you'll become ineligible.

**Comment by vnwhlr** [comment]
Also, maybe we ought to standardize the method name we export - for example, if all modules export a function called something like `computeEligibility`, we could treat them homogeneously when we get to a point where we need to compute eligibility for multiple benefits at once.

**Comment by isaacchansky** [comment]
I like the idea of exporting a common function. Also super +1 on the 'reasons'. and idea of how close you are to eligibility would also be really cool. 
Along with 'reasons' I think an 'asOf' timestamp could be nice, which would be static and based off of whatever policy used was written.

**Comment by vnwhlr** [comment]
Wiki article here: https://github.com/codeforboston/project-hope/wiki/Benefit-Eligibility-Modules-Interface

**Comment by geedo0** [comment]
For the sake of the front-end, it would be really good to programmatically expose the inputs to the various eligibility modules. I'm thinking something like a schema for input validation and some metadata like a form label.

**Comment by rastapopolous** [comment]
@geedo0 are you imagining an object w properties all the stored values on the client, that can then be passed to a given module?  that sounds to me like the sensible approach...

**Comment by geedo0** [comment]
@rastapopolous Yes, and in constructing the form to build this client object. We can make a call to get_parameters on each module to get a complete list of things to ask for. Or we can dynamically create forms for one particular module.

**Comment by eanyanwu** [comment]
Evan has started work on this based on our discussion last week. Still under works, but the basic signature is there. We can hash this out more this Tuesday.

https://github.com/codeforboston/project-hope/wiki/Benefit-Eligibility-Modules-Interface

**Comment by knod** [comment]
I'm pretty homebrewed, so haven't worked with others on architecture much, but another option is a benefits program object that has a `.get_eligibility()` function and a `.form_inputs` object or array that looks a bit like this:

```js
program.form_inputs = [
  { name: 'age', form_type: 'number', min: 0 },
  { name: 'gross_income', form_type: 'number', min: 0 }
];
```

I'd gotten the impression that with js, you can take advantage of objects and their methods/properties instead of using getters and setters. For that matter, we could make the `form_parameters` object more useful in that it could contain the form information that's sent back. Something like:

```js
program.form_inputs = [
  { name: 'Age', form_type: 'number', min: 0, user_value: '45' },
  { name: 'Gross monthly income', form_type: 'number', min: 0, user_value: null },
  { name: 'Housing agency median income requirements', form_type: 'radio', options: [ 30, 50, 80 ], user_value: null }
];
```

Some of the values we're looking for may not be client properties, so this might be a useful way to get at that kind of information. For example, the housing agency requirements that I've found in the Section 8 Housing Voucher, which the case worker would probably put in (let me know if you have questions about why that is, but there are reasons).

It may also allow us to pre-fill the form with data we already have about the client.

I have some further variations on this running around in my head as I write this that could be geared more towards our html, and other, needs.

**Comment by knod** [comment]
Alternatively, that more complex `form_inputs` object could be constructed separately elsewhere and the `form_parameters` array could just be a way to refer to which properties of that object are needed.
